### ISSUE TITLE
google_secret_manager_secret_iam_policy forces replacement with no changes

### ISSUE THREAD

**Issue by adamallred** [terraform-providers/terraform-provider-google#5963]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request.
* Please do not leave _+1_ or _me too_ comments, they generate extra noise for issue followers and do not help prioritize the request.
* If you are interested in working on this issue or have submitted a pull request, please leave a comment.
* If an issue is assigned to the `modular-magician` user, it is either in the process of being autogenerated, or is planned to be autogenerated soon. If an issue is assigned to a user, that user is claiming responsibility for the issue. If an issue is assigned to `hashibot`, a community member has claimed the issue already.

### Terraform Version

Terraform/0.12.20 TFE/v202002-2
Google provider: 3.14.0

### Affected Resource(s)

* google_secret_manager_secret_iam_policy 

### Terraform Configuration Files

```tf
provider "google" {
  version = "3.14.0"
}

provider "google-beta" {
  version = "3.14.0"
}

resource "google_project" "test-secret-iam" {
  name            = "<my-project-name>"
  project_id      = "<my-project-id>"
  folder_id       = "<my-folder-id>"
  billing_account = "<my-billing-account>"
}

# define an IAM policy that provides secret access.
data "google_iam_policy" "secret-access" {
  binding {
    role = "roles/secretmanager.secretAccessor"
    members = [
      "user:<my-user>"
    ]
  }
  binding {
    role = "roles/secretmanager.viewer"
    members = [
      "user:<my-user>"
    ]
  }
}

resource "google_secret_manager_secret" "my-secret" {
  provider  = google-beta
  project   = google_project.test-secret-iam.project_id
  secret_id = "secret-key"
  replication {
    automatic = true
  }
  depends_on = [google_project_service.secretmanager]
}

resource "google_secret_manager_secret_version" "my-secret-version" {
  provider    = google-beta
  secret      = google_secret_manager_secret.my-secret.name
  secret_data = "supersecret"
}

resource "google_secret_manager_secret_iam_policy" "my-secret-access" {
  provider    = google-beta
  project     = google_secret_manager_secret.my-secret.project
  secret_id   = google_secret_manager_secret.my-secret.name
  policy_data = data.google_iam_policy.secret-access.policy_data
}

resource "google_project_service" "secretmanager" {
  service            = "secretmanager.googleapis.com"
  project            = google_project.test-secret-iam.project_id
  disable_on_destroy = false
}

```

### Debug Output

https://gist.github.com/adamallred/27202caac850b384eac3484734ffe278

### Expected Behavior

After applying the changes above, re-running a `plan` with no changes would lead to no changes being detected.
 
### Actual Behavior

After making the above resources, re-running a `plan` with no changes leads to an IAM policy replacement being required due to changed project value from the project number to the project ID.

If I update the HCL to refer to:

```tf
resource "google_secret_manager_secret_iam_policy" "my-secret-access" {
  provider    = google-beta
  project     = google_project.test-secret-iam.number
  secret_id   = google_secret_manager_secret.my-secret.name
  policy_data = data.google_iam_policy.secret-access.policy_data
}
``` 

then the plan sees no changes.

I'm not sure if this behavior is expected due to the authoritative nature of `google_secret_manager_secret_iam_policy` (maybe it always forces a replace to be authoritative), but being able to set `.number` to achieve no changes seems like that would circumvent that behavior. 

### Steps to Reproduce

1. Apply the HCL above
2. Re-plan with no changes, and observe forced replacement.
3. Change HCL to use the snippet just above, and see no planned changes.

**Comment by edwardmedia** [comment]
I do see below `plan` right after `tf apply`
```

Terraform will perform the following actions:
  # google_secret_manager_secret_iam_policy.my-secret-access must be replaced
-/+ resource "google_secret_manager_secret_iam_policy" "my-secret-access" {
      ~ etag        = "BwWhoW/L//4=" -> (known after apply)
      ~ id          = "projects/89056095xxxx/secrets/secret-key" -> (known after apply)
        policy_data = jsonencode(
            {
                bindings = [
                    {
                        members = [
                            "user:abc@abc.com",
                        ]
                        role    = "roles/secretmanager.secretAccessor"
                    },
                    {
                        members = [
                            "user:abc@abc.com",
                        ]
                        role    = "roles/secretmanager.viewer"
                    },
                ]
            }
        )
      ~ project     = "89056095xxxx" -> "myproject" # forces replacement
        secret_id   = "projects/89056095xxxx/secrets/secret-key"
    }
Plan: 1 to add, 0 to change, 1 to destroy.
```

**Comment by slevenick** [comment]
Hm, yeah that's an issue.

I think the best way to avoid this would be to change your config to use:
```
resource "google_secret_manager_secret_iam_policy" "my-secret-access" {
  provider    = google-beta
  project     = google_secret_manager_secret.my-secret.project
  secret_id   = google_secret_manager_secret.my-secret.secret_id
  policy_data = data.google_iam_policy.secret-access.policy_data
}
```
instead. This changes `secret_id = google_secret_manager_secret.my-secret.name` => `secret_id = google_secret_manager_secret.my-secret.secret_id`. *Or* use `.id` instead of `.name`

The weird thing here is that `google_secret_manager_secret.name` is actually `projects/{project_number}/secrets/{secret_id}`, and the IAM policy attempts to parse the input to figure out what project it belongs to, finding the project number instead of the project ID

I'll investigate a fix

**Comment by slevenick** [comment]
I would alternatively recommend not setting the `project` field if you use `google_secret_manager_secret.my-secret.name`. The reason it causes issues is that you are specifying two different forms of `project` within the config by setting both `project` and using `google_secret_manager_secret.my-secret.name` as the `secret_id`.

**Comment by adamallred** [comment]
Ah, I missed that I used `name` instead of `secret_id`, the latter being what's present in the docs. I'll check that `secret_id` behaves as expected.

**Comment by adamallred** [comment]
Using `secret_id` works as expected.

I don't know if you want to continue to investigate the behavior when `.name` is used or not, I'll leave it to you to keep the issue open or close it. From my end, I'm all good.

Thanks!

**Comment by slevenick** [comment]
I looked into creating a plan-time error for this (and similar issues) where the project doesn't match a long name's project. The issue that I ran into in this case is that `name` is computed, so we cannot in any circumstance get a plan time error from depending on the value of it. The plan time validation actually made the issue *worse* because it would allow creation of the policy with non matching projects, and then prevented deletion.

Closing this for now as there isn't a way to have reliable plan time validation on this, I'll add a note to the docs
